src/test/lib/DataTest.cpp

```diff
@@ -3556,6 +3556,248 @@ void QuicTestOperationPriority()
+            MsQuicStream Stream3(*Connections[2], QUIC_STREAM_OPEN_FLAG_UNIDIRECTIONAL, CleanUpManual, ConnectionPriorityTestContext::ClientStreamStartStreamCallback, &Context);
+            Context.ExpectedStream = &Stream1;
+
+            Stream1.Start(QUIC_STREAM_START_FLAG_IMMEDIATE);
+            // Wait until this StreamStart operation is drained
+            TEST_TRUE(Context.BlockAfterInitialStart.WaitTimeout(TestWaitTimeout));
+            Context.ExpectedStream = &Stream3;
+
+            for (uint8_t i = 0; i < ConnectionPriorityTestContext::NumSend; ++i) {
+                Streams[i] = new(std::nothrow) MsQuicStream(*Connections[1], QUIC_STREAM_OPEN_FLAG_UNIDIRECTIONAL, CleanUpManual, ConnectionPriorityTestContext::ClientStreamStartStreamCallback, &Context);
+                TEST_QUIC_SUCCEEDED(Streams[i]->GetInitStatus());
+                TEST_QUIC_SUCCEEDED(Streams[i]->Send(&Buffer, 1, QUIC_SEND_FLAG_START | QUIC_SEND_FLAG_FIN));
+            }
+
+            TEST_QUIC_SUCCEEDED(Stream3.Start(QUIC_STREAM_START_FLAG_PRIORITY_WORK));
+            TEST_QUIC_SUCCEEDED(Stream3.Send(&Buffer, 1, QUIC_SEND_FLAG_FIN | QUIC_SEND_FLAG_PRIORITY_WORK));
+            Context.OperationQueuedComplete.Set(); // All operations are queued. Kick off processing the operations
+
+            MsQuicStream* ExpectedStartOrder[ConnectionPriorityTestContext::NumSend + 2] = {0};
+            ExpectedStartOrder[0] = &Stream1;
+            ExpectedStartOrder[1] = &Stream3;
+            for (uint8_t i = 2; i < ConnectionPriorityTestContext::NumSend + 2; ++i) {
+                ExpectedStartOrder[i] = Streams[i-2];
+            }
+            MsQuicStream* ExpectedSendOrder[ConnectionPriorityTestContext::NumSend + 1] = {0};
+            ExpectedSendOrder[0] = &Stream3;
+            for (uint8_t i = 1; i < ConnectionPriorityTestContext::NumSend + 1; ++i) {
+                ExpectedSendOrder[i] = Streams[i-1];
+            }
+
+            TEST_TRUE(Context.AllReceivesComplete.WaitTimeout(TestWaitTimeout));
+
+            TEST_TRUE(memcmp(Context.StartOrder, ExpectedStartOrder, sizeof(ExpectedStartOrder)) == 0);
+            TEST_TRUE(memcmp(Context.SendOrder, ExpectedSendOrder, sizeof(ExpectedSendOrder)) == 0);
+            TEST_TRUE(Context.TestSucceeded);
+            for (uint8_t i = 0; i < ConnectionPriorityTestContext::NumSend; ++i) {
+                delete Streams[i];
+            }
+        }
+
+        // processing                                                             | queued
+        // [1[s0]]                                                                | []
+        // [1[s0]]                                                                | [2[s(n-1)0p, s(n-1)2p, s00, s02, ..., s(n-2)0, s(n-2)2]]
+        // [1[s0]]                                                                | [2[s(n-1)0p, s(n-1)2p, s00, s02, ..., s(n-2)0, s(n-2)2], 3[s0p, s2p]]
+        // [2[s(n-1)0p, s(n-1)2p, s00, s02, ..., s(n-2)0, s(n-2)2], 3[s0p, s2p]]  | []
```

Comment:
for some reason only `2[s(n-2)2]` is processed after `3[s2p]`. Bug? or expected?

Reply:
OK simply callback call is delayed. Splitting StreamStart/StreamSend can resolve.